proposals/0000-unsatisfiable.rst

```diff
@@ -0,0 +1,599 @@
+
+     class Eq a => ReflexiveEq a where
+         reflexiveEq :: a -> a -> Bool
+         reflexiveEq = (==)
+
+     instance Unsatisfiable (Text "Can't compare functions with reflexiveEq") => ReflexiveEq (a -> b)
+
+     type DoubleMsg = Text "Equality is not reflexive on Double"
+     instance Unsatisfiable DoubleMsg => ReflexiveEq Double where
+         reflexiveEq = unsatisfiable @DoubleMsg
+
+   Even though ``Eq`` is a superclass of ``ReflexiveEq``, the instance does not
+   need to list it in the context, because the Given ``Unsatisfiable``
+   constraint suffices to solve the ``Eq (a -> b)`` constraint.  This means that
+   use sites will not accidentally duplicate error messages as with the current
+   behaviour of ``TypeError``.
+
+   Moreover, when ``-fdefer-type-errors`` is used to call ``reflexiveEq 0 (0 ::
+   Double)``, this will result in a runtime exception that correctly blames the
+   use of ``-fdefer-type-errors`` to bypass the ``Unsatisfiable`` constraint.
+
+#. The following is regarded as total by the pattern-match coverage checker
+   (thanks to `/u/ComicIronic on Reddit
+   <https://www.reddit.com/r/haskell/comments/p2ao7v/unsatisfiable_a_ghcproposal_for_better_custom/h8je78s/>`_
+   for the example)::
+
+     data MyGADT a where
+       MyInt :: MyGADT Int
+
+     type family IsBool a where
+       IsBool Bool = ()
+       IsBool a    = Unsatisfiable (Text "Must be Bool")
+
+     foo :: IsBool a => MyGADT a -> Void
+     foo x = case x of {}
+
+   This means it is possible to use ``Unsatisfiable`` to get custom error
+   messages, without needing to write cases that are in practice inaccessible.
+   Similar functionality was recently implemented for ``TypeError`` by Sam
+   Derbyshire (see `#20180
+   <https://gitlab.haskell.org/ghc/ghc/-/issues/20180>`_).
+
+#. The following is accepted despite the apparent functional dependency violation::
+
+     class C a b | a -> b
+     instance Unsatisfiable (Text "No") => C a b
+
+   That is, an instance can be ruled out with a custom type error even where
+   this would otherwise conflict with the functional dependencies.  A practical
+   use case for this `arises in the optics library
+   <https://hackage.haskell.org/package/optics-core-0.4/docs/Optics-Internal-Optic-Subtyping.html#t:JoinKinds>`_.
+   Without this rule, a workaround is possible by building an otherwise unused
+   cycle in the context, but this runs the risk of exposing the cycle as a
+   constraint solver loop::
+
+     instance (Unsatisfiable (Text "No"), C a b) => C a b
+
+#. Assuming the ``Warning`` option is included, the following ``decode``
+   function will emit a warning if it is used at type ``Integer``, but the
+   program will otherwise compile normally::
+
+     type family WarnInteger a where
+       WarnInteger Integer = Warning (Text "Integer may require unbounded memory!")
+       WarnInteger a       = ()
+
+     decode :: (FromJSON a, WarnInteger a) => ByteString -> Maybe a
```

Comment:
I like this. There are a bunch of instances like this out there that should be used quite carefully but that are all too easy to use without realising it. We could decorate them with such a warning.